### ISSUE TITLE
No ARM64 Container

### ISSUE THREAD

**Issue by nniehoff** [promlabs/promlens-public#11]
I'd like to be able to run promlens on a raspberry pi, which would require an ARM64 container, hopefully you can just build and push an arm64 container and there isn't much change required.

**Comment by juliusv** [comment]
@nniehoff Thanks for the suggestion! I'm taking a look now at how Prometheus does this (https://github.com/prometheus/prometheus/pull/5031/files), and it seems it needs a QEMU-based solution on Linux (https://www.ecliptik.com/Cross-Building-and-Running-Multi-Arch-Docker-Images/) to build images for foreign architectures. OTOH since Go itself can already do crossbuilding, maybe it's sufficient to just use that and then put the resulting binaries into arch-specific busybox / alpine images. In any case, will try to get to this soon!

**Comment by mhausenblas** [comment]
Did you consider using https://docs.docker.com/buildx/working-with-buildx/ already?

**Comment by juliusv** [comment]
Yeah, I saw that, but got distracted :/ Installed the latest Docker engine with buildx support now and trying it out! I also have to release a new PromLens version anyway (new PromQL features), so hopefully I can already support multi-arch images for that one :)

**Comment by juliusv** [comment]
Hmm, I'm getting cgo compilation errors for the SQLite3 package on ARM with buildx, but it doesn't show more detail on what's actually going wrong (even with `go build -x`). In case I cannot figure that out, I could try disabling SQLite3 on ARM. Is either one of you using it?

**Comment by juliusv** [comment]
(this would only be for storing shared links, but there's GCS and MySQL options for that too)

**Comment by juliusv** [comment]
This is all the info it's giving me when trying to directly build the package on `linux/arm64/v8` with buildx:

```
 => ERROR [builder 5/6] RUN go build -x github.com/mattn/go-sqlite3                                                                                                                                                                                                         5.1s
------                                                                                                                                                                                                                                                                           
 > [builder 5/6] RUN go build -x github.com/mattn/go-sqlite3:                                                                                                                                                                                                                    
#14 1.409 WORK=/tmp/go-build891035528                                                                                                                                                                                                                                            
#14 3.583 mkdir -p $WORK/b001/                                                                                                                                                                                                                                                   
#14 3.587 cd /go/pkg/mod/github.com/mattn/go-sqlite3@v1.14.4                                                                                                                                                                                                                     
#14 3.587 TERM='dumb' CGO_LDFLAGS='"-g" "-O2" "-ldl"' /usr/local/go/pkg/tool/linux_arm64/cgo -objdir $WORK/b001/ -importpath github.com/mattn/go-sqlite3 -- -I $WORK/b001/ -g -O2 -std=gnu99 -DSQLITE_ENABLE_RTREE -DSQLITE_THREADSAFE=1 -DHAVE_USLEEP=1 -DSQLITE_ENABLE_FTS3 -DSQLITE_ENABLE_FTS3_PARENTHESIS -DSQLITE_ENABLE_FTS4_UNICODE61 -DSQLITE_TRACE_SIZE_LIMIT=15 -DSQLITE_OMIT_DEPRECATED -DSQLITE_DISABLE_INTRINSIC -DSQLITE_DEFAULT_WAL_SYNCHRONOUS=1 -DSQLITE_ENABLE_UPDATE_DELETE_LIMIT -Wno-deprecated-declarations -DHAVE_PREAD64=1 -DHAVE_PWRITE64=1 -I/go/pkg/mod/github.com/mattn/go-sqlite3@v1.14.4 ./backup.go ./callback.go ./error.go ./sqlite3.go ./sqlite3_context.go ./sqlite3_load_extension.go ./sqlite3_opt_userauth_omit.go ./sqlite3_other.go ./sqlite3_type.go
#14 4.902 go build github.com/mattn/go-sqlite3: /usr/local/go/pkg/tool/linux_arm64/cgo: exit status 2
```

**Comment by mhausenblas** [comment]
Interesting.

> Is either one of you using it?

I am not.

>  trying to directly build the package on linux/arm64/v8 with buildx:

Did you try others, like `arm/v7` or `armhf`? Do they result in the same error?

**Comment by juliusv** [comment]
> Did you try others, like arm/v7 or armhf? Do they result in the same error?

It fails similarly obscurely in a cgo step, but not 100% the same (exit code 1 and not 2)...

You can reproduce it with the following minimal Dockerfile:

```Dockerfile
FROM golang:1.17
RUN go get -x github.com/mattn/go-sqlite3
```

...and then:

```bash
# These fail.
docker buildx build --platform linux/arm64/v8 -f ./Dockerfile ./
docker buildx build --platform linux/arm/v7 -f ./Dockerfile ./
docker buildx build --platform linux/armhf -f ./Dockerfile ./

# This works fine.
docker buildx build --platform linux/amd64 -f ./Dockerfile ./
```

In case anyone wants to help figure things out :)

**Comment by mhausenblas** [comment]
Gotcha. Stupid question, but rather than `go get` how about `go install`?

**Comment by juliusv** [comment]
That just gives me `package github.com/mattn/go-sqlite3 is not a main package`, but other than that I assume it would run into the same build issues if it was a main package...

**Comment by kinduff** [comment]
I was able to build this on a Raspberry Pi 3. [Full output here](https://gist.github.com/kinduff/0fdd1b63f7137c66bdc67d0b4294ee61).

How can I test this further? I have other devices where this can be tested.

As far as I'm seeing the issue could be with `buildx`, any more information I'll be happy to assist since I have experience doing builds for multiple platforms.

**Comment by tonistiigi** [comment]
> You can reproduce it with the following minimal Dockerfile:

Seems to work fine for me(from intel with --platform=linux/arm64), assuming you are on Docker Desktop with emulators preinstalled or have used https://github.com/tonistiigi/binfmt manually.

A better way would be to actually cross-compile and not rely on qemu.

```
FROM --platform=$BUILDPLATFORM tonistiigi/xx:1.1.0 AS xx

FROM --platform=$BUILDPLATFORM golang:1.17 AS build
COPY --from=xx / /
ARG TARGETPLATFORM
# I usually prefer clang based builds but keep it minimal here
RUN xx-apt-get install -y --no-install-recommends gcc
RUN xx-go get -x  github.com/mattn/go-sqlite3

FROM scratch
# replace path with binary you are actually building
COPY --from=build /path/to/binary .
```

The performance difference for the `go get` command is 6m20s in qemu vs 7s for xx.

More reading: 
- https://github.com/tonistiigi/xx
- https://medium.com/@tonistiigi/faster-multi-platform-builds-dockerfile-cross-compilation-guide-part-1-ec087c719eaf

**Comment by juliusv** [comment]
> I was able to build this on a Raspberry Pi 3.

@kinduff Interesting, thanks! So the issue is with my qemu-based Docker buildx environment on Linux then I guess... I'll need to get that fixed :)

**Comment by juliusv** [comment]
@tonistiigi So I am on Ubuntu + amd64 on the host, and after fixing my binfmt stuff for qemu, my buildx driver supposedly supports a bunch of archs:

```
suspicious_murdock *  docker-container                    
  suspicious_murdock0 unix:///var/run/docker.sock running linux/amd64, linux/arm64, linux/ppc64le, linux/s390x, linux/386, linux/mips64le, linux/mips64, linux/arm/v7, linux/arm/v6
```

However, with that, now I get SEGFAULT errors like:

```
$ docker buildx build --platform linux/arm/v7 -f ./Dockerfile_bug.buildx ./
WARN[0000] No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load 
[+] Building 4.7s (6/6) FINISHED                                                                                                       
 => [internal] load build definition from Dockerfile_bug.buildx                                                                   0.1s
 => => transferring dockerfile: 114B                                                                                              0.0s
 => [internal] load .dockerignore                                                                                                 0.1s
 => => transferring context: 2B                                                                                                   0.0s
 => [internal] load metadata for docker.io/library/golang:1.17                                                                    3.1s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                     0.0s
 => CACHED [1/2] FROM docker.io/library/golang:1.17@sha256:eb61213fe612b6af346cab13c2b81f1b8113f9ccf23a5ca6b54fede8ffd63bc7       0.0s
 => => resolve docker.io/library/golang:1.17@sha256:eb61213fe612b6af346cab13c2b81f1b8113f9ccf23a5ca6b54fede8ffd63bc7              0.0s
 => ERROR [2/2] RUN go get -x github.com/mattn/go-sqlite3@latest                                                                  1.4s
------                                                                                                                                 
 > [2/2] RUN go get -x github.com/mattn/go-sqlite3@latest:                                                                             
#6 0.311 qemu: Unsupported syscall: 403                                                                                                
#6 0.398 # get https://proxy.golang.org/github.com/mattn/@v/list                                                                       
#6 0.399 # get https://proxy.golang.org/github.com/mattn/go-sqlite3/@v/list                                                            
#6 0.400 # get https://proxy.golang.org/github.com/@v/list                                                                             
#6 1.011 qemu: uncaught target signal 11 (Segmentation fault) - core dumped
#6 1.102 Segmentation fault (core dumped)
------
Dockerfile_bug.buildx:2
--------------------
   1 |     FROM golang:1.17
   2 | >>> RUN go get -x github.com/mattn/go-sqlite3@latest
   3 |     
--------------------
error: failed to solve: process "/bin/sh -c go get -x github.com/mattn/go-sqlite3@latest" did not complete successfully: exit code: 139
```

...but it varies. Sometimes I also get SIGILL (illegal instruction) errors like this:

```
docker buildx build --platform linux/arm/v7 -f ./Dockerfile_bug.buildx ./
WARN[0000] No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load 
[+] Building 1.6s (5/5) FINISHED                                                                                                       
 => [internal] load build definition from Dockerfile_bug.buildx                                                                   0.1s
 => => transferring dockerfile: 114B                                                                                              0.0s
 => [internal] load .dockerignore                                                                                                 0.1s
 => => transferring context: 2B                                                                                                   0.0s
 => [internal] load metadata for docker.io/library/golang:1.17                                                                    0.3s
 => CACHED [1/2] FROM docker.io/library/golang:1.17@sha256:eb61213fe612b6af346cab13c2b81f1b8113f9ccf23a5ca6b54fede8ffd63bc7       0.0s
 => => resolve docker.io/library/golang:1.17@sha256:eb61213fe612b6af346cab13c2b81f1b8113f9ccf23a5ca6b54fede8ffd63bc7              0.0s
 => ERROR [2/2] RUN go get -x github.com/mattn/go-sqlite3@latest                                                                  1.0s
------                                                                                                                                 
 > [2/2] RUN go get -x github.com/mattn/go-sqlite3@latest:                                                                             
#5 0.284 qemu: Unsupported syscall: 403                                                                                                
#5 0.367 # get https://proxy.golang.org/github.com/@v/list                                                                             
#5 0.368 # get https://proxy.golang.org/github.com/mattn/go-sqlite3/@v/list                                                            
#5 0.369 # get https://proxy.golang.org/github.com/mattn/@v/list                                                                       
#5 0.640 qemu: uncaught target signal 4 (Illegal instruction) - core dumped
#5 0.733 Illegal instruction (core dumped)
------
Dockerfile_bug.buildx:2
--------------------
   1 |     FROM golang:1.17
   2 | >>> RUN go get -x github.com/mattn/go-sqlite3@latest
   3 |     
--------------------
error: failed to solve: process "/bin/sh -c go get -x github.com/mattn/go-sqlite3@latest" did not complete successfully: exit code: 132
```

But I think I may have to check out and understand all your other links and tips soon, because right now I have little idea what I'm actually doing :)

**Comment by tonistiigi** [comment]
🤔 `403` should be `clock_gettime64(2)`. Afaics it is implemented since qemu 5.0 and in docker seccomp policy from 20.10.0 so not quite sure what you are missing here. Make sure you use `tonistiigi/binfmt` image though, the similar distro packages are known to cause issues.

**Comment by jpetazzo** [comment]
@juliusv I think you're on the right track, and I suspect that the issues that you're seeing are none of your doing, but probably weird bugs between QEMU and the particular CPU that you're using. That's just a gut feeling though.

For what it's worth, I was able to compile your test case:
```
$ cat Dockerfile 
FROM golang:1.17
RUN go get -x github.com/mattn/go-sqlite3
```

```
$ docker buildx build --platform linux/arm64/v8 .
WARN[0000] No output specified for docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load 
[+] Building 265.9s (6/6) FINISHED                                                                                                                                                                       
 => [internal] load build definition from Dockerfile                                                                                                                                                0.0s
 => => transferring dockerfile: 96B                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                   0.0s
 => => transferring context: 2B                                                                                                                                                                     0.0s
 => [internal] load metadata for docker.io/library/golang:1.17                                                                                                                                      2.1s
 => [auth] library/golang:pull token for registry-1.docker.io                                                                                                                                       0.0s
 => [1/2] FROM docker.io/library/golang:1.17@sha256:eb61213fe612b6af346cab13c2b81f1b8113f9ccf23a5ca6b54fede8ffd63bc7                                                                               26.1s
 => => resolve docker.io/library/golang:1.17@sha256:eb61213fe612b6af346cab13c2b81f1b8113f9ccf23a5ca6b54fede8ffd63bc7                                                                                0.0s
 => => sha256:d19c6eaaced152f2555856673c7fddada222c210af0071877297faa8944a5ebf 126B / 126B                                                                                                          0.1s
 => => sha256:387334821c5766fe1e51787a0e39fabd91c968e763b23a2444bb011742679aa3 80.99MB / 80.99MB                                                                                                   22.3s
 => => sha256:7df682714696dcda51e3d1450e19e12e91596cb835720e25c9c4b9e1c4bf4c16 102.63MB / 102.63MB                                                                                                 24.3s
 => => sha256:ee42e24e52cb2f9b50ab14328865298f5ae6ac159219db890afc6366ec641a46 54.67MB / 54.67MB                                                                                                   14.9s
 => => sha256:aa8c15f93136056df893ba8088992a38e90dc37210fd6ebb712413fd045c58de 10.66MB / 10.66MB                                                                                                    3.8s
 => => sha256:96565d527970b1cc982493f96d6b317b1fbfdd7bf0ea6fd58d44942bb4f597c5 4.94MB / 4.94MB                                                                                                      1.0s
 => => sha256:6476cb406b24c94f24348db05d154b814b14ee38860dea1625080e7e08295a80 53.62MB / 53.62MB                                                                                                    7.8s
 => => extracting sha256:6476cb406b24c94f24348db05d154b814b14ee38860dea1625080e7e08295a80                                                                                                           0.8s
 => => extracting sha256:96565d527970b1cc982493f96d6b317b1fbfdd7bf0ea6fd58d44942bb4f597c5                                                                                                           0.1s
 => => extracting sha256:aa8c15f93136056df893ba8088992a38e90dc37210fd6ebb712413fd045c58de                                                                                                           0.1s
 => => extracting sha256:ee42e24e52cb2f9b50ab14328865298f5ae6ac159219db890afc6366ec641a46                                                                                                           1.0s
 => => extracting sha256:387334821c5766fe1e51787a0e39fabd91c968e763b23a2444bb011742679aa3                                                                                                           1.0s
 => => extracting sha256:7df682714696dcda51e3d1450e19e12e91596cb835720e25c9c4b9e1c4bf4c16                                                                                                           1.7s
 => => extracting sha256:d19c6eaaced152f2555856673c7fddada222c210af0071877297faa8944a5ebf                                                                                                           0.0s
 => [2/2] RUN go get -x github.com/mattn/go-sqlite3                                                                                                                                               237.2s
```

This is the recipe that I use to enable QEMU+binfmt:
```
docker run --rm --privileged aptman/qus -s -- -p
```

And for what it's worth, my CPU is a `Intel(R) Core(TM) i5-9600KF CPU @ 3.70GHz`.

(Just providing one data point hoping that helps!)

**Comment by juliusv** [comment]
Sweet! Thanks so much everyone, then it does look like my qemu setup is just broken (since I'm using distro packages), and perhaps I can even get it to work without qemu using `xx` as @tonistiigi suggested. Hoping to get back to this in the next days and then release an ARM version of PromLens if everything goes well :)

**Comment by juliusv** [comment]
Alright, finally got around to diving into all of this deeper and read your blog post, @tonistiigi, thanks for that and all the tooling! And thanks to @jpetazzo for pointers to an ARM VM to test on :) I got it working now, so planning a release with this in the upcoming days.

Btw. I also still had to fiddle a bit with Go's external linker flags to produce a fully static build that would work both with the cross-compilation and run in the target image (Alpine in my case).

**Comment by tonistiigi** [comment]
> that would work both with the cross-compilation and run in the target image (Alpine in my case).

Just FYI that xx does work fine in Alpine as well. You just need to install clang as there are no prebuilt gcc variants. Examples of that are in the readme and in example projects. You may still want to do static binaries of course but there is no reason to use Debian-based stage only for a cross-compile build if you will use the results in alpine anyway.

**Comment by juliusv** [comment]
@tonistiigi Good point, I think I like that even better and switched everything to Alpine and non-static binaries - thanks :)

**Comment by juliusv** [comment]
Thanks again everyone, just published PromLens v1.1.0 with ARM images, which I managed to test on an ARM VM :)
